system/include/compat/avxintrin.h

```diff
@@ -44,215 +535,1271 @@ _mm_broadcast_ss(const float *__mem_addr)
+  __extension__({                                                              \
+    (index) < 4 ? _mm_extract_epi32((__a).v0, (index))                         \
+                : _mm_extract_epi32((__a).v1, (index) - 4);                    \
+  })
+
+#define _mm256_extract_epi16(X, N)                                             \
+  __extension__({                                                              \
+    (N) < 8 ? _mm_extract_epi16((X).v0, (N))                                   \
+            : _mm_extract_epi16((X).v1, (N) - 8);                              \
+  })
+
+#define _mm256_extract_epi8(X, N)                                              \
+  __extension__({                                                              \
+    (N) < 16 ? _mm_extract_epi8((X).v0, (N))                                   \
+             : _mm_extract_epi8((X).v1, (N) - 16);                             \
+  })
+
+#define _mm256_extract_epi64(X, N)                                             \
+  __extension__({                                                              \
+    (N) < 2 ? _mm_extract_epi64((X).v0, (N))                                   \
+            : _mm_extract_epi64((X).v1, (N) - 2);                              \
+  })
+
+#define _mm256_insert_epi32(X, I, N)                                           \
+  __extension__({                                                              \
+    ((N) & 7) < 4                                                              \
+      ? _mm256_set_m128i((X).v1, _mm_insert_epi32((X).v0, (I), (N) & 3))       \
+      : _mm256_set_m128i(_mm_insert_epi32((X).v1, (I), (N) & 3), (X).v0);      \
+  })
+
+#define _mm256_insert_epi16(X, I, N)                                           \
+  ({                                                                           \
+    ((N) & 0xF) < 8                                                            \
+      ? _mm256_set_m128i((X).v1, _mm_insert_epi16((X).v0, (I), (N) & 0x7))     \
+      : _mm256_set_m128i(_mm_insert_epi16((X).v1, (I), (N) & 0x7), (X).v0);    \
+  })
+
+#define _mm256_insert_epi8(X, I, N)                                            \
+  ({                                                                           \
+    ((N) & 0x1F) < 16                                                          \
+      ? _mm256_set_m128i((X).v1, _mm_insert_epi8((X).v0, (I), (N) & 0xF))      \
+      : _mm256_set_m128i(_mm_insert_epi8((X).v1, (I), (N) & 0xF), (X).v0);     \
+  })
+
+#define _mm256_insert_epi64(X, I, N)                                           \
+  ({                                                                           \
+    ((N) & 3) < 2                                                              \
+      ? _mm256_set_m128i((X).v1, _mm_insert_epi64((X).v0, (I), (N) & 1))       \
+      : _mm256_set_m128i(_mm_insert_epi64((X).v1, (I), (N) & 1), (X).v0);      \
+  })
```

Comment:
I wonder if it would be better to use the `__attribute__((__always_inline__, __nodebug__))` style of function representation for these? Or was there a particular reason to have these as defines, but some of the previous functions as individual functions?

(iirc the cmp was special cased to make it forced-obvious to the compiler to inline and discard dead code on the compare branches that aren't taken, but I think that `__attribute__((__always_inline__))` would achieve the same effect.

Reply:
Yes, it's special, because parameter of  some intrinsic need to be compile time constant,  constexpr in C++ is suitable for this, but in this header should be work both in C and C++,  so some intrinsic is written in macro.

e.g. wasm_i32x4_extract_lane, the __i must be __builtin_constant_p().

static __inline__ int32_t __DEFAULT_FN_ATTRS wasm_i32x4_extract_lane(v128_t __a,
                                                                       int __i)
      __REQUIRE_CONSTANT(__i) {
    return ((__i32x4)__a)[__i];
  }

Reply:
Gotcha, that makes sense.